roles/app-data/tasks/main.yml

```diff
@@ -7,16 +7,29 @@
 
   when: app_data is not defined
 
+- name: Create our temporary add_data_container
+  set_fact:
+    app_data_container: []
+
 - block:
-    - name: Merge our appdata with any custom declarations in app_instance_data
+    - name: Merge app_data into app_data_container
       set_fact:
-        app_data: "{{ app_data|default([])|merge_custom_app_data(app_instance_data|default([])) }}"
+        app_data: "{{ app_data_container | merge_custom_app_data( app_data | default([])) }}"
+
+- block:
+    - name: Merge app_instance_data into app_data_container overwriting app_data definitions
+      set_fact:
+        app_data: "{{ app_data_container | merge_custom_app_data( app_instance_data | default([])) }}"
 
   when: app_instance_data is defined
 
 - block:
-    - name: Merge our appdata with any custom declarations in custom_app_instance_data
+    - name: Merge custom_app_instance_data into app_data_container overwriting app_data and app_instance_data definitions
       set_fact:
-        app_data: "{{ app_data|default([])|merge_custom_app_data(custom_app_instance_data|default([])) }}"
+        app_data: "{{ app_data_container | merge_custom_app_data(custom_app_instance_data | default([])) }}"
```

Comment:
you shouldn't have the default here on `custom_app_instance_data` as the `when` makes sure it is defined.

perhaps
`when: custom_app_instance_data|default([])|count > 0` would be a better check